metriq_gym/resource_estimation.py

```diff
@@ -0,0 +1,436 @@
+    import random
+
+    random.seed(params.seed)
+
+    num_qubits = params.num_qubits
+    graph_type: GraphType = params.graph_type
+    qaoa_layers = params.qaoa_layers
+    circuit_encoding: EncodingType = "Direct"
+
+    if device is None:
+        raise ValueError("LR-QAOA benchmark requires a device to estimate resources.")
+
+    if device.id == "aer_simulator" and graph_type == "NL":
+        graph_device = rx.generators.star_graph(num_qubits)
+    else:
+        graph_device = connectivity_graph(device)
+
+    edges_device = list(graph_device.edge_list())
+
+    if graph_type == "1D":
+        edges = [(i, i + 1) for i in range(num_qubits - 1)]
+    elif graph_type == "NL":
+        if graph_device.num_nodes() != num_qubits:
+            raise ValueError("Number of qubits does not match the device connectivity graph.")
+        edges = [(edge[0], edge[1]) for edge in edges_device]
+    elif graph_type == "FC":
+        edges = [(i, j) for i in range(num_qubits) for j in range(i + 1, num_qubits)]
+        device_edges = {(edge[0], edge[1]) for edge in edges_device}
+        if not device_edges.issuperset(edges):
+            circuit_encoding = "SWAP"
+    else:
+        raise ValueError(
+            f"Unsupported graph type: {graph_type}. Supported types are '1D', 'NL', and 'FC'."
+        )
+
+    possible_weights = [0.1, 0.2, 0.3, 0.5, 1.0]
+    graph = nx.Graph()
+    graph.add_nodes_from(range(num_qubits))
+    graph_info = [(i, j, random.choice(possible_weights)) for i, j in edges]
+    graph.add_weighted_edges_from(graph_info)
+
+    circuits = prepare_qaoa_circuit(
+        graph=graph,
+        qaoa_layers=qaoa_layers,
+        graph_type=graph_type,
+        circuit_encoding=circuit_encoding,
+    )
+
+    circuits_with_params: list[QuantumCircuit] = []
+    for _ in range(params.trials):
+        for p_layer, circuit in zip(qaoa_layers, circuits):
+            linear_ramp = list(range(1, p_layer + 1))
+            betas = [i * params.delta_beta / p_layer for i in reversed(linear_ramp)]
+            gammas = [i * params.delta_gamma / p_layer for i in linear_ramp]
+            circuits_with_params.append(circuit.assign_parameters(betas + gammas))
+
+    return [CircuitBatch(circuits=circuits_with_params, shots=params.shots)]
+
+
+EstimatorFunction = Callable[[BaseModel, QuantumDevice | None], list[CircuitBatch]]
+
+
+ESTIMATORS: dict[JobType, EstimatorFunction] = {
+    JobType.WIT: _estimate_wit,
+    JobType.QML_KERNEL: _estimate_qml_kernel,
+    JobType.QUANTUM_VOLUME: _estimate_quantum_volume,
+    JobType.BSEQ: _estimate_bseq,
+    JobType.CLOPS: _estimate_clops,
+    JobType.MIRROR_CIRCUITS: _estimate_mirror_circuits,
+    JobType.BERNSTEIN_VAZIRANI: _estimate_qedc,
+    JobType.PHASE_ESTIMATION: _estimate_qedc,
+    JobType.HIDDEN_SHIFT: _estimate_qedc,
+    JobType.QUANTUM_FOURIER_TRANSFORM: _estimate_qedc,
+    JobType.LR_QAOA: _estimate_lr_qaoa,
```

Comment:
Those estimation functions should be responsibility of the benchmark (files in benchmarks.py)